x/wasm/internal/keeper/genesis_test.go

```diff
@@ -67,6 +68,166 @@ func TestGenesisExportImport(t *testing.T) {
 	require.False(t, dstIT.Valid())
 }
 
+func TestFailFastImport(t *testing.T) {
+	wasmCode, err := ioutil.ReadFile("./testdata/contract.wasm")
+	require.NoError(t, err)
+	codeHash := sha256.Sum256(wasmCode)
+	anyAddress := make([]byte, 20)
+
+	specs := map[string]struct {
+		src        types.GenesisState
+		expSuccess bool
+	}{
+		"happy path: code info correct": {
+			src: types.GenesisState{
+				Codes: []types.Code{{
+					CodeInfo: wasmTypes.CodeInfo{
+						CodeHash: codeHash[:],
+						Creator:  anyAddress,
+					},
+					CodesBytes: wasmCode,
+				}},
+				Contracts: nil,
+			},
+			expSuccess: true,
+		},
+		"prevent code hash mismatch": {src: types.GenesisState{
+			Codes: []types.Code{{
+				CodeInfo: wasmTypes.CodeInfo{
+					CodeHash: make([]byte, len(codeHash)),
+					Creator:  anyAddress,
+				},
+				CodesBytes: wasmCode,
+			}},
+			Contracts: nil,
+		}},
+		"happy path: code info and contract do match": {
+			src: types.GenesisState{
+				Codes: []types.Code{{
+					CodeInfo: wasmTypes.CodeInfo{
+						CodeHash: codeHash[:],
+						Creator:  anyAddress,
+					},
+					CodesBytes: wasmCode,
+				}},
+				Contracts: []types.Contract{
+					{
+						ContractAddress: addrFromUint64(1<<32 + 1),
+						ContractInfo: wasmTypes.ContractInfo{
+							CodeID:  1,
+							Creator: anyAddress,
+							Label:   "any",
+							Created: &types.AbsoluteTxPosition{BlockHeight: 1, TxIndex: 1},
+						},
+					},
+				},
+			},
+			expSuccess: true,
+		},
+		"prevent contracts that points to non existing codeID": {
+			src: types.GenesisState{
+				Contracts: []types.Contract{
+					{
+						ContractAddress: addrFromUint64(1<<32 + 1),
+						ContractInfo: wasmTypes.ContractInfo{
+							CodeID:  1,
+							Creator: anyAddress,
+							Label:   "any",
+							Created: &types.AbsoluteTxPosition{BlockHeight: 1, TxIndex: 1},
+						},
+					},
+				},
+			},
+		},
+		"prevent duplicate contracts": {
```

Comment:
:+1:

Reply:
Also good to show we do allow two contracts with different contractAddresses to point to the same CodeID